src/python/pants/engine/native.py

```diff
@@ -286,6 +287,25 @@ def extern_get_type_for(self, context_handle, val):
     type_id = c.to_id(type(obj))
     return TypeId(type_id)
 
+  @_extern_decl('Handle', ['ExternContext*', 'TypeId'])
+  def extern_get_handle_from_type_id(self, context_handle, ty):
+    c = self._ffi.from_handle(context_handle)
+    return c.to_value(ty)
+
+  @_extern_decl('TypeId', ['ExternContext*', 'TypeId'])
+  def extern_get_union_for(self, context_handle, type_id):
+    """Return an optional union if the type belongs to one"""
+
+    c = self._ffi.from_handle(context_handle)
+    input_type = c.from_id(type_id.tup_0)
+
+    union_rules = self._build_configuration.union_rules()
+
+    if input_type in union_rules:
+      return TypeId(c.to_id(type(input_type)))
+    else:
+      return TypeId(0)
```

Comment:
Can simplify to below.

```python
      return TypeId(c.to_id(type(input_type))) if input_type in union_rules else TypeId(0)
```

Not everyone agrees this is an improvement and you can ignore this if you'd like.